### ISSUE TITLE
istio-testing failing to update some branches

### ISSUE THREAD

**Issue by fejta** [istio/test-infra#1511]
Looks like the job cannot update istio/community and istio-ecosystem/istio-coredns-plugin:

/cc @geeknoid 

```console

{"component":"branchprotector","error":"update istio: update community: update master from protected=true: get current branch protection: getting branch protection 404: Not Found","file":"prow/cmd/branchprotector/protect.go:127","func":"main.main","level":"error","msg":"0","time":"2019-07-18T18:22:49Z"}
{"component":"branchprotector","error":"update istio-ecosystem: update istio-coredns-plugin: update master from protected=true: get current branch protection: getting branch protection 404: Not Found","file":"prow/cmd/branchprotector/protect.go:127","func":"main.main","level":"error","msg":"1","time":"2019-07-18T18:22:49Z"}
{"component":"branchprotector","file":"prow/cmd/branchprotector/protect.go:129","func":"main.main","level":"fatal","msg":"Encountered 2 errors protecting branches","time":"2019-07-18T18:22:49Z"}
```

**Comment by cjwagner** [comment]
The branchprotector cronjob is still failing with errors like the following:
```yaml
{
  component:  "branchprotector"   
  error:  "update istio-ecosystem: update istio-coredns-plugin: update master from protected=true: get current branch protection: getting branch protection 404: Not Found"   
  file:  "prow/cmd/branchprotector/protect.go:127"   
  func:  "main.main"   
  level:  "error"   
  msg:  "0"   
}
{
  component:  "branchprotector"   
  error:  "update istio: update community: update thisisnotapril-patch-2 from protected=false: get policy: required prow jobs require branch protection"   
  file:  "prow/cmd/branchprotector/protect.go:127"   
  func:  "main.main"   
  level:  "error"   
  msg:  "1"   
}
{
  component:  "branchprotector"   
  error:  "update istio: update community: update hklai-testrelease-group from protected=false: get policy: required prow jobs require branch protection"   
  file:  "prow/cmd/branchprotector/protect.go:127"   
  func:  "main.main"   
  level:  "error"   
  msg:  "1"   
}

**Comment by geeknoid** [comment]
I deleted some stale branches just now on the istio/community repo which should eliminate the two errors coming from that repo.

**Comment by geeknoid** [comment]
For the istio-ecosystem errors, I sent an invitation from istio-ecosystem to have istio-bot join that org. Once the invitation is accepted, that bot can be made an admin and thus will be able to do branch-protection stuff.

Someone needs to accept the GitHub invitation for istio-bot though.

**Comment by geeknoid** [comment]
OK, so deleting those branches in the istio/community made it so things go further. Branch protection on the bots repo is now being applied correctly in particular.

**Comment by geeknoid** [comment]
Dupe of #1422?

**Comment by geeknoid** [comment]
I  added the bot to the istio-ecosystem org, so hopefully that issue is gone and the bot is working properly.

Can we put alerting in place somewhere to point out when the bot has failed?

**Comment by cjwagner** [comment]
> Can we put alerting in place somewhere to point out when the bot has failed?

This is tracked as part of https://github.com/istio/test-infra/issues/1396. When we switch the branch protector from a k8s cronjob to a ProwJob we can use the same alerting mechanisms we use for other ProwJobs (e.g. Testgrid email alerts).

**Comment by geeknoid** [comment]
The  branch protector seems finicky, can we make it a bit more robust? In particular, could it continue applying branch protection to other repo/branches even if one repo/branch fails? As it is, a single failure prevents anything after it from being suitably protected.

Check out this log entry that is the current point of failure. The error says that branches controlled  by  prow must be protected. The error is being reported against a private developer-created branch on the istio/istio.io repo.

[link](https://pantheon.corp.google.com/logs/viewer?interval=NO_LIMIT&project=istio-testing&minLogLevel=0&expandAll=false&timestamp=2019-08-03T04%3A34%3A42.110000000Z&customFacets&limitCustomFacetWidth=true&advancedFilter=resource.type%3D%22container%22%0Aresource.labels.pod_id%3D%22branchprotector-1564806600-dc48x%22%0Aresource.labels.zone%3D%22us-west1-a%22%0Aresource.labels.project_id%3D%22istio-testing%22%0Aresource.labels.cluster_name%3D%22prow%22%0Aresource.labels.container_name%3D%22branchprotector%22%0Aresource.labels.namespace_id%3D%22default%22%0Aresource.labels.instance_id%3D%223242663893665764462%22%0Atimestamp%3D%222019-08-03T04%3A33%3A53.000000000Z%22%0AinsertId%3D%22nv6mabfd7e0oj%22&scrollTimestamp=2019-08-03T04%3A33%3A53.000000000Z
)

**Comment by geeknoid** [comment]
I'm reopening this since the issue is still happening. At the present moment, branch protection settings specified in config.yaml are not applied to all repos they should be due to the branchprotector encountering errors in one repo and bailing for the rest.

**Comment by geeknoid** [comment]
So I'd like to see three things:

- BrandhProtector should continue to run even when it encounters an error on a single repo. So a single broken repo doesn't mean that all subsequent repos are also broken.

- Trigger an alert when a failure occurs so we know we need to fix things.

- Trigger branch protector to run as soon as a PR happens to config.yaml rather than waiting for the cron job.

**Comment by howardjohn** [comment]
2 and 3 can probably be handled by converting to a prow job so that we can put it on the testgrid (alerts) and as a postsubmit (happen immediately)

**Comment by clarketm** [comment]
- [x] `BranchProtector` should continue to run even when it encounters an error on a single repo. So a single broken repo doesn't mean that all subsequent repos are also broken. kubernetes/test-infra#14231
- [x] Trigger an alert when a failure occurs so we know we need to fix things (Now that #1770 is merged, testgrid annotations can be added as desired).
- [x] Trigger branch protector to run as soon as a PR happens to config.yaml rather than waiting for the cron job. #1770
